--- a/drivers/spi/spidev.c
+++ b/drivers/spi/spidev.c
@@ -712,6 +712,7 @@
 	{ .name = "m53cpld" },
 	{ .name = "spi-petra" },
 	{ .name = "spi-authenta" },
+	{ .name = "spidev" },	
 	{ .name = "em3581" },
 	{ .name = "si3210" },
 	{},
@@ -740,6 +741,7 @@
 	{ .compatible = "micron,spi-authenta", .data = &spidev_of_check },
 	{ .compatible = "rohm,dh2228fv", .data = &spidev_of_check },
 	{ .compatible = "semtech,sx1301", .data = &spidev_of_check },
+	{ .compatible = "linux,spidev", .data = &spidev_of_check },
 	{ .compatible = "silabs,em3581", .data = &spidev_of_check },
 	{ .compatible = "silabs,si3210", .data = &spidev_of_check },
 	{},
