--- a/net/eth.c	2016-08-01 12:11:47.954123318 +0200
+++ b/net/eth.c	2016-07-29 12:49:29.906622000 +0200
@@ -184,8 +184,8 @@
 		printf("\nWarning: %s using MAC address from net device\n",
 			dev->name);
 	} else if (!(eth_address_set(dev->enetaddr))) {
-		printf("\nError: %s address not set.\n",
-		       dev->name);
+		//printf("\nError: %s address not set.\n",
+		//     dev->name);
 		return -EINVAL;
 	}
 
