--- a/include/configs/sama5d3_xplained.h
+++ b/include/configs/sama5d3_xplained.h
@@ -4,6 +4,8 @@
  * Copyright (C) 2014 Atmel Corporation
  *		      Bo Shen <voice.shen@atmel.com>
  *
+ * Modified for the THOR NX-T1 board 2016-08-01 by Roy Schneider.
+ *
  * SPDX-License-Identifier:	GPL-2.0+
  */
 
@@ -43,6 +45,15 @@
 	(CONFIG_SYS_SDRAM_BASE + 4 * 1024 - GENERATED_GBL_DATA_SIZE)
 #endif
 
+/* SPI for boot picture */
+#define CONFIG_ATMEL_SPI
+#define CONFIG_ATMEL_HLCD
+#define CONFIG_LCD
+#define CONFIG_LCD_INFO
+#define LCD_BPP LCD_COLOR16
+#define LCD_OUTPUT_BPP 24
+#define CONFIG_SYS_WHITE_ON_BLACK
+
 /* NAND flash */
 #define CONFIG_CMD_NAND
 
@@ -81,7 +92,7 @@
 #define CONFIG_PHYLIB
 
 /* MMC */
-#define CONFIG_CMD_MMC
+/*#define CONFIG_CMD_MMC*/
 
 #ifdef CONFIG_CMD_MMC
 #define CONFIG_MMC
@@ -91,7 +102,7 @@
 #endif
 
 /* USB */
-#define CONFIG_CMD_USB
+/*#define CONFIG_CMD_USB*/
 
 #ifdef CONFIG_CMD_USB
 #define CONFIG_USB_ATMEL
