--- a/include/configs/sama5d3_xplained.h
+++ b/include/configs/sama5d3_xplained.h
@@ -12,6 +12,12 @@
 #include <linux/sizes.h>
 #include "at91-sama5_common.h"
 
+#ifndef CONFIG_CONSOLE_SCROLL_LINES
+  #define CONFIG_CONSOLE_SCROLL_LINES 5
+#endif
+
+#define LIFT_APP_THOR_NXT_GENERATION  2  // Define wheter the Bootloader is made for the NX-T3 (3) or NX-T2 (2) generation boards.
+
 /*
  * This needs to be defined for the OHCI code to work but it is defined as
  * ATMEL_ID_UHPHS in the CPU specific header files.
@@ -27,6 +33,13 @@
 #define CFG_SYS_SDRAM_BASE           0x20000000
 #define CFG_SYS_SDRAM_SIZE		0x10000000
 
+/* SPI/LCD for boot picture */
+//#define CONFIG_ATMEL_HLCD
+//#define CONFIG_LCD
+//#define CONFIG_LCD_INFO
+#define LCD_BPP 4 // 16 bit colour depth
+#define LCD_OUTPUT_BPP 24
+
 /* NAND flash */
 #ifdef CONFIG_CMD_NAND
 #define CFG_SYS_NAND_BASE		0x60000000
@@ -40,6 +53,9 @@
 
 /* size of u-boot.bin to load */
 
+#define CONFIG_SYS_NAND_U_BOOT_OFFS	0x10000
+#define CONFIG_SYS_NAND_U_BOOT_SIZE	0xB0000
+
 /* Falcon boot support on raw MMC */
 /* U-Boot proper stored by default at 0x200 (256 KiB) */
 
