--- a/include/configs/sama5d3_xplained.h
+++ b/include/configs/sama5d3_xplained.h
@@ -4,6 +4,8 @@
  * Copyright (C) 2014 Atmel Corporation
  *		      Bo Shen <voice.shen@atmel.com>
  *
+ * Modified for the THOR NX-T1 board 2016-08-01 by Roy Schneider.
+ *
  * SPDX-License-Identifier:	GPL-2.0+
  */
 
@@ -12,6 +14,10 @@
 
 #include "at91-sama5_common.h"
 
+#ifndef CONFIG_CONSOLE_SCROLL_LINES
+  #define CONFIG_CONSOLE_SCROLL_LINES 5
+#endif
+
 /*
  * This needs to be defined for the OHCI code to work but it is defined as
  * ATMEL_ID_UHPHS in the CPU specific header files.
@@ -35,6 +41,15 @@
 	(CONFIG_SYS_SDRAM_BASE + 16 * 1024 - GENERATED_GBL_DATA_SIZE)
 #endif
 
+/* SPI for boot picture */
+#define CONFIG_ATMEL_SPI
+#define CONFIG_ATMEL_HLCD
+#define CONFIG_LCD
+#define CONFIG_LCD_INFO
+#define LCD_BPP LCD_COLOR16
+#define LCD_OUTPUT_BPP 24
+#define CONFIG_SYS_WHITE_ON_BLACK
+
 /* NAND flash */
 #define CONFIG_CMD_NAND
 
@@ -108,7 +123,8 @@
 #elif CONFIG_SYS_USE_NANDFLASH
 #define CONFIG_SPL_NAND_DRIVERS
 #define CONFIG_SPL_NAND_BASE
-#define CONFIG_SYS_NAND_U_BOOT_OFFS	0x40000
+#define CONFIG_SYS_NAND_U_BOOT_OFFS	0x20000
+#define CONFIG_SYS_NAND_U_BOOT_SIZE	0xA0000
 #define CONFIG_SYS_NAND_5_ADDR_CYCLE
 #define CONFIG_SYS_NAND_PAGE_SIZE	0x800
 #define CONFIG_SYS_NAND_PAGE_COUNT	64
