--- a/include/configs/sama5d3_xplained.h
+++ b/include/configs/sama5d3_xplained.h
@@ -2,6 +2,8 @@
 /*
  * Configuration settings for the SAMA5D3 Xplained board.
  *
+ * Modified for the THOR NX-T2/3 board 2021-03-31 by Roy Schneider.
+ *
  * Copyright (C) 2014 Atmel Corporation
  *		      Bo Shen <voice.shen@atmel.com>
  */
@@ -11,6 +13,12 @@
 
 #include "at91-sama5_common.h"
 
+#ifndef CONFIG_CONSOLE_SCROLL_LINES
+  #define CONFIG_CONSOLE_SCROLL_LINES 5
+#endif
+
+#define LIFT_APP_THOR_NXT_GENERATION  2  // Define that the Bootloader is made for the NX-T3 generation boards.
+
 /*
  * This needs to be defined for the OHCI code to work but it is defined as
  * ATMEL_ID_UHPHS in the CPU specific header files.
@@ -33,6 +39,15 @@
 	(CONFIG_SYS_SDRAM_BASE + 16 * 1024 - GENERATED_GBL_DATA_SIZE)
 #endif
 
+/* SPI/LCD for boot picture */
+#define CONFIG_ATMEL_SPI
+#define CONFIG_ATMEL_HLCD
+#define CONFIG_LCD
+#define CONFIG_LCD_INFO
+#define LCD_BPP LCD_COLOR16
+#define LCD_OUTPUT_BPP 24
+//#define CONFIG_SYS_WHITE_ON_BLACK
+
 /* NAND flash */
 #ifdef CONFIG_CMD_NAND
 #define CONFIG_SYS_MAX_NAND_DEVICE	1
@@ -75,7 +90,8 @@
 #define CONFIG_SPL_NAND_DRIVERS
 #define CONFIG_SPL_NAND_BASE
 #endif
-#define CONFIG_SYS_NAND_U_BOOT_OFFS	0x40000
+#define CONFIG_SYS_NAND_U_BOOT_OFFS	0x10000
+#define CONFIG_SYS_NAND_U_BOOT_SIZE	0xB0000
 #define CONFIG_SYS_NAND_5_ADDR_CYCLE
 #define CONFIG_SYS_NAND_PAGE_SIZE	0x800
 #define CONFIG_SYS_NAND_PAGE_COUNT	64
